### ISSUE TITLE
`object-shorthand`: add option to warn on explicit return arrows

### ISSUE THREAD

**Issue by ljharb** [eslint/eslint#7564]
I think I clicked on the right issue template:

**Tell us about your environment**

* **ESLint Version:** 3.9.1
* **Node Version:** 7
* **npm Version:** stock 3.x

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

```
"object-shorthand": "error"
```

**What did you do? Please include the actual source code causing the issue.**

```js
const obj = {
  foo: () => {},
  bar: (x, y) => { const something = foo(); return something(bar); },
};
```

**What did you expect to happen?**
The rule to prefer:
```js
const obj = {
  foo() {},
  bar(x, y) { const something = foo(); return something(bar); },
};
```

**What actually happened? Please include the actual, raw output from ESLint.**
No warning.

Basically, I'd like to add an option to `object-shorthand` that prefers concise (named) methods over explicit-return arrow functions.

**Comment by not-an-aardvark** [comment]
This seems reasonable to me, I've wanted something like this as well.

However, what would the option do in this case?

```js
const obj = {
  foo: () => {
    return this.bar;
  }
};
``` 

Using object shorthand here would yield a different result.

Also, would the rule have a preference between these two cases?

```js
// Case A
const obj = {
  foo: () => 5
};

// Case B
const obj2 = {
  foo() {
    return 5;
  }
};
```

I ask the second question because it seems like the result of autofixing would be unclear for this case:

```js
/* eslint arrow-body-style: 2, object-shorthand: [2, "theNewOption"] */

const obj = {
  foo: () => {
    return 5;
  }
};
```

If the `arrow-body-style` fixer acted first, the resulting code would look like Case A, but if the `object-shorthand` fixer acted first, the resulting code would look like Case B. This might not necessarily be a problem, but we should think about that as we consider the scope of this option.

**Comment by ljharb** [comment]
@not-an-aardvark good point! Clearly it would have to ignore methods *and* arrow functions that contain `this`, since otherwise the refactor could change the meaning.

I do agree there could be a priority race condition between `object-shorthand` and `arrow-body-style` - it's worth calling out, but I don't see it as a conflict (assuming the first issue you mentioned is rectified), just as a source of nondeterminism. Is there a way to prioritize one rule's fix over another's, or to control ordering of autofixers?

**Comment by platinumazure** [comment]
@ljharb There isn't a way to prioritize a rule's fix (fixes are applied in reverse order of location in the code, basically). I think one way to fix the issue @not-an-aardvark brought up is to add an option to `arrow-body-style` to avoid reporting/fixing object methods that use arrow functions, which users could use to avoid conflicts in this area.

**Comment by not-an-aardvark** [comment]
In terms of deterministic autofixing, it would be easier if this new `object-shorthand` option disallowed all arrow functions as properties, not just explicit-return functions. But I can see why that might be undesirable.

**Comment by ljharb** [comment]
Yeah that's definitely undesirable in most cases, which is why I want to restrict it to explicit return (or empty function bodies).

**Comment by not-an-aardvark** [comment]
I'll champion this proposal.

**Comment by not-an-aardvark** [comment]
@eslint/eslint-team Is anyone willing to :+1: this proposal? We need two more :+1:s from the team to accept it.

**Comment by btmills** [comment]
Just added the third :+1:, marking as accepted!

**Comment by not-an-aardvark** [comment]
We never ended up deciding on an option name. I'm thinking it should be something along the lines of `avoidLongformArrows`, but I'm open to bikeshedding on that. (Normally I would prefer `allowLongformArrows` instead of `avoidLongformArrows`, but the rule already has an `avoidQuotes` option, so maybe we should stay consistent with that.)

**Comment by ljharb** [comment]
The terminology used to describe them are "explicit return arrow functions", i've never seen "long form" before. Perhaps for a name, `avoidExplicitReturnArrows`?

I agree that if you've set the rule to "never", this other rule doesn't really make sense.

**Comment by not-an-aardvark** [comment]
After looking at the rule docs again, it seems like `avoidQuotes` is actually doing something unrelated, so I think it would be fine to have a positively-worded option. How about `allowExplicitReturnArrows`?

**Comment by ljharb** [comment]
The default would have to remain "no enforcement" tho, and having an option where absence and `false` aren't the same is weird. Maybe `enforceOnExplicitReturnArrows`, or something else where `true` enables it?

**Comment by not-an-aardvark** [comment]
Personally I don't think having an option default to `true` is confusing, but I'm willing to change it if it's confusing to others.

**Comment by not-an-aardvark** [comment]
Let's move the discussion to https://github.com/eslint/eslint/pull/7746.

**Comment by nzakas** [comment]
Question: why are arrow functions with an explicit return special? (I noted it was said avoiding all arrow functions was undesirable, but no reason was given.)

**Comment by ljharb** [comment]
Explicit return arrows are not special - when they are not using `this`, they are trivially similar to concise methods - i.e., they can contain statements. Implicit return arrows are the things that are special, in that they are functions that can only contain one expression and implicitly return its value.

**Comment by nzakas** [comment]
@ljharb sorry, I think I wasn't clear. I don't understand why it makes sense to avoid only one form of arrow function for this option.  I can understand wanting to prevent arrow functions from being assigned to properties, but saying the presence of "return" in the body is the determining factor is the type of overly specific targeting that we've tried to avoid in core rules. Otherwise, we will end up with requests for each type of overly specific condition people might have. 

I definitely support having an option to prefer methods over arrow functions (as that's my preference), just the specific nature of this proposal seems problematic to me.

**Comment by not-an-aardvark** [comment]
Just to make sure we're on the same page, the rule isn't checking for the presence of `return`, it's checking for a `BlockStatement` as the arrow function body.

So this would also get reported:

```js
var x = {
  foo: () => {
    bar();
    // no return statement
  }
};
```

But this would not get reported:

```js
var x = {
  foo: () => bar() // body is a CallExpression instead of a BlockStatement
};
```

**Comment by ljharb** [comment]
Right - the point is to prefer the more concise form when it's effectively equivalent, which is only the case in the block body form (called "explicit return" even when it omits `return`). With the implicit form, the concise method version would be more verbose and would require adding a `return`, and so I wouldn't want to encourage users to avoid arrows in that case - ie, I want only implicit-form arrow functions used in object literals, because they're special.
